### PR TITLE
Improve persistent volume claim validation

### PR DESCRIPTION
**Pull request by derekwaynecarr**
This PR does the following: 
- If a pvc.spec.volumeName has been assigned, do not allow the spec.accessModes or spec.resources to be change
- Adds new tests to validate pvc update

/cc @markturansky @kubernetes/rh-storage @kubernetes/goog-cluster

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -905,8 +905,26 @@ func ValidatePersistentVolumeClaim(pvc *api.PersistentVolumeClaim) field.ErrorLi
 }
 
 func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *api.PersistentVolumeClaim) field.ErrorList {
-	allErrs := field.ErrorList{}
-	allErrs = ValidatePersistentVolumeClaim(newPvc)
+	allErrs := ValidateObjectMetaUpdate(&newPvc.ObjectMeta, &oldPvc.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidatePersistentVolumeClaim(newPvc)...)
+	// if a pvc had a binded volume, we should not allow updates to resources or access modes
```
Comment (markturansky) [inline pkg/api/validation/validation.go]: s/binded/bound

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -905,8 +905,26 @@ func ValidatePersistentVolumeClaim(pvc *api.PersistentVolumeClaim) field.ErrorLi
...
+	if len(oldPvc.Spec.VolumeName) != 0 {
+		specPath := field.NewPath("spec")
+		newAccessModes := sets.NewString()
+		for _, accessMode := range newPvc.Spec.AccessModes {
```
Comment (markturansky) [inline pkg/api/validation/validation.go]: A helper exists to make a deterministic string from access modes.

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/helpers.go#L203

**Comment by markturansky** [comment]
A couple little things but otherwise LGTM.

**Comment by derekwaynecarr** [comment]
Thanks @markturanksy, will update and poke

**Comment by derekwaynecarr** [comment]
Pushed updates, thanks for pointers.

**Comment by markturansky** [comment]
LGTM.

**Comment by markturansky** [comment]
@derekwaynecarr since this hasn't merged yet, would you mind updating the validation such that the entire spec cannot be mutated after binding?

We use the deletion of the claim as a trigger for reclaiming the resource.  Unbinding the claim by removing the volume name will get fixed through the controller, but it doesn't need to be possible for the end user anyway. The entire spec is ignored post-binding (except to set volumeName as needed), so there's no reason to not lock the whole thing. Once volumeName is set, deletion of the claim is the only way to unbind them.

**Comment by derekwaynecarr** [comment]
@markturansky - made requested update.

**Comment by markturansky** [comment]
Thanks @derekwaynecarr LGTM

**Pull request merged**